cpp/src/arrow/csv/writer.cc

```diff
@@ -390,7 +407,7 @@ class CSVWriterImpl : public ipc::RecordBatchWriter {
     for (int col = 0; col < schema_->num_fields(); col++) {
       const std::string& col_name = schema_->field(col)->name();
       header_length += col_name.size();
-      header_length += CountEscapes(col_name);
+      header_length += options_.escaping ? CountEscapes(col_name) : 0;
```

Comment:
The same goes here, if escaping is disabled, we should ensure that no escaping is actually required.

Reply:
(We should also cover these with tests.)